image-rs/src/signature/mechanism/simple/xrss.rs

```diff
@@ -0,0 +1,167 @@
+    mechanism::simple,
+};
+
+#[derive(Debug, Default)]
+pub struct RegistryClient {
+    // reqwest client for the container registry that supports the X-R-S-S extension
+    pub client: Client,
+}
+
+#[derive(Deserialize, Debug, PartialEq, Eq, Serialize, Default)]
+pub struct GetOAuthTokenResponse {
+    #[serde(default)]
+    pub token: String,
+    #[serde(default)]
+    pub access_token: String,
+    #[serde(default)]
+    pub expires_in: u32,
+    #[serde(default)]
+    pub issued_at: String,
+}
+
+#[derive(Deserialize, Debug, PartialEq, Eq, Serialize, Default)]
+pub struct GetSignaturesResponse {
+    #[serde(default)]
+    signatures: Vec<RegistrySignature>,
+}
+
+#[derive(Deserialize, Debug, PartialEq, Eq, Serialize, Default)]
+pub struct RegistrySignature {
+    #[serde(default, rename = "schemaVersion")]
+    pub schema_version: u32,
+    #[serde(default)]
+    pub name: String,
+    #[serde(default, rename = "type")]
+    pub signature_type: String,
+    #[serde(default)]
+    pub content: String,
+}
+
+impl RegistryClient {
+    pub fn new() -> RegistryClient {
+        RegistryClient {
+            client: reqwest::Client::new(),
+        }
+    }
+
+    async fn get_oauth_token_for_registry(
+        &self,
+        image: &Image,
+        auth: &RegistryAuth,
+    ) -> Result<String> {
+        match auth {
+            RegistryAuth::Anonymous => {
+                bail!("Trying to get signature from a registry without providing auth information")
+            }
+            RegistryAuth::Basic(username, password) => {
+                let scope = format!("repository:{}:pull", image.reference.repository());
+                let res = self
+                    .client
+                    .get(format!(
+                        "https://{}/oauth/token",
+                        image.reference.registry()
+                    ))
+                    .query(&[
+                        ("account", username),
+                        ("scope", &scope),
+                        ("service", &"registry".to_string()),
+                    ])
+                    .basic_auth(username, Some(password))
+                    .send()
+                    .await
+                    .map_err(|e| anyhow!("Failed to fetch oauth token for registry: {:?}", e))?;
+
+                let oauth_token_response_body: GetOAuthTokenResponse =
+                    res.json::<GetOAuthTokenResponse>().await.map_err(|e| {
+                        anyhow!(
+                            "Unexpected response from fetching oauth token from registry: {:?}",
+                            e
+                        )
+                    })?;
+
+                Ok(oauth_token_response_body.token)
+            }
+        }
+    }
+
+    pub async fn get_signatures_from_registry(
+        &self,
+        image: &Image,
+        digest: &Digest,
+        auth: &RegistryAuth,
+    ) -> Result<Vec<Vec<u8>>> {
+        let mut sigs: Vec<Vec<u8>> = simple::new_sigs_vec().await;
```

Comment:
Also, why not directly use
```rust
let mut sigs: Vec<Vec<u8>> = Vec::new();
```

Reply:
I made this change and the code builds OK.